pkg/kubectl/rolling_updater.go

```diff
@@ -93,27 +79,436 @@ const (
+		readinessCheck: updater.pollForReadyPods,
+		createdCheck:   updater.getCreatedPods,
+	}
+	return updater
+}
+
+// Update all pods for a ReplicationController (oldRc) by creating a new
+// controller (newRc) with 0 replicas, and synchronously scaling oldRc and
+// newRc until oldRc has 0 replicas and newRc has the original # of desired
+// replicas. Cleanup occurs based on a RollingUpdaterCleanupPolicy.
+//
+// The scaling increment each interval is either 1 or based on a percent of
+// the desired replicas. The default scaling direction is up/down. If
+// percentage is negative, the direction is down/up for in-place updates.
+//
+// When a percentage is used, the updater will compute a minimum pod readiness
+// requirement and ensure that a minimum number of pods will be ready for the
+// duration of the update. Each interval, the updater will scale down whatever
+// it can without violating the minimum, and will scale up as much as it can
+// up to a maximum increment. This means amount scaled up or down each
+// interval will vary based on the timeliness of readiness and the updater
+// will always try to make progress, even slowly.
+//
+// If an update from newRc to oldRc is already in progress, we attempt to
+// drive it to completion. If an error occurs at any step of the update, the
+// error will be returned.
 //
-// After the successful scale-up of the controller, the controller is given to
-// the UpdateAcceptor. If the UpdateAcceptor rejects the controller, the
-// deployment is stopped with an error.
-type UpdateAcceptor interface {
-	// Accept returns nil if the controller is okay, otherwise returns an error.
-	Accept(*api.ReplicationController) error
+// TODO: make this handle performing a rollback of a partially completed
+// rollout.
+func (r *RollingUpdater) Update(config *RollingUpdaterConfig) error {
+	out := config.Out
+	oldRc := config.OldRc
+
+	// Find an existing controller (for continuing an interrupted update) or
+	// create a new one if necessary.
+	sourceId := fmt.Sprintf("%s:%s", oldRc.Name, oldRc.UID)
+	newRc, existed, err := r.getOrCreateTargetController(config.NewRc, sourceId)
+	if err != nil {
+		return err
+	}
+	if existed {
+		fmt.Fprintf(out, "Continuing update with existing controller %s.\n", newRc.Name)
+	} else {
+		fmt.Fprintf(out, "Created %s\n", newRc.Name)
+	}
+	// Extract the desired replica count from the controller.
+	desired, err := strconv.Atoi(newRc.Annotations[desiredReplicasAnnotation])
+	if err != nil {
+		return fmt.Errorf("Unable to parse annotation for %s: %s=%s",
+			newRc.Name, desiredReplicasAnnotation, newRc.Annotations[desiredReplicasAnnotation])
+	}
+
+	// This is the maximum number of replicas we'll scale in either direction.
+	maxIncrement := 1
+	// This is the minimum number of replicas (total of old/new) which must
+	// remain ready for the duration of the update if progress is to be made.
+	minReady := oldRc.Spec.Replicas - maxIncrement
```

Comment:
Thinking about this:

`min(oldRc.originalSpec.replicas, newRc.spec.replicas) - maxIncrement`

Shouldn't it be:

`min(oldRc.originalSpec.replicas, newDesired) - maxIncrement`

The new replicas may be 0 initially, and in any case the minimum should be the same across continuations.

So as you described, if we don't have access to the original oldRc replicas to support that calculation, I'm not sure continutations are going to look exactly right all the time even with symmetrical updates, because newRc.Replicas+oldRc.Replicas isn't always going to be equal to the desired count (or the original oldrc replica count) in the case of a partial scale due to a scale failure, timeout, or other interruption.

Even though `AbortRollingUpdate` is broken for asymmetric updates, we don't actually reject asymmetric controller pairs... I guess if we want to defer the old replica count tracking we'd have to accept the behavior you outlined in https://github.com/GoogleCloudPlatform/kubernetes/pull/11942#discussion_r36012184 and use:

`newRc.Spec.Replicas + oldRc.Spec.Replicas - maxIncrement`

If you're rather go with that I can rip out the current asymmetric test cases I have and we can do followup work to correctly support those cases.